### PR TITLE
Prevent from sleeping

### PR DESCRIPTION
**Pull request by jbouaziz**
Instead of preventing the mac from going to sleep, we use a preference and only prevent app nap if the device is connected to a power source.

See https://github.com/mipstian/catch/issues/24

### CODE REVIEW

**Comment by mipstian** [comment]
Thanks!

I double-checked and I have two comments:
- the power assertions are incorrect. We always want to prevent app nap, but only prevent system sleep if the checkbox is checked.
- why refresh the power assertion status when the preferences window is opened? Opening a window shouldn't change that. Maybe that was test code?

**Comment by mipstian** [comment]
Oh it's not when the window is opened. But still, a UI refresh method shouldn't call into power stuff :)

**Comment by jbouaziz** [comment]
Do you mind commenting directly on the lines you're talking about please?

File: CTCPreferencesController.m [hunk 1]
```diff
@@ -46,6 +50,8 @@ - (void)refreshInvalidInputMarkers {
     NSImage *invalidImage = [NSImage imageNamed:@"warning"];
     self.torrentsSavePathWarningImageView.image = CTCDefaults.isTorrentsSavePathValid ? validImage : invalidImage;
     self.feedURLWarningImageView.image = CTCDefaults.isFeedURLValid ? validImage : invalidImage;
+    
+    [CTCScheduler.sharedScheduler updateAppNapStatus];
```
Comment (mipstian) [inline CTCPreferencesController.m]: The point of this method is to update the UI (namely, the warning markers), so it shouldn't be calling into code that changes the power assertions.

File: CTCScheduler.m [hunk 2]
```diff
@@ -91,19 +94,29 @@ - (void)preventAppNap {
...
+  if ([NSProcessInfo.processInfo respondsToSelector:@selector(beginActivityWithOptions:reason:)]
+      && self.activityToken) {
       [NSProcessInfo.processInfo endActivity:self.activityToken];
       self.activityToken = nil;
   }
 }
 
 - (void)updateAppNapStatus {
-    self.polling ? [self preventAppNap] : [self allowAppNap];
+    [self shouldPreventAppNap] ? [self preventAppNap] : [self allowAppNap];
```
Comment (mipstian) [inline CTCScheduler.m]: Here you're switching between preventing app nap + system sleep, and preventing neither. What we want is always prevent app nap, and only conditionally prevent system sleep.

**Comment by dombarnes** [comment]
Is this PR going to be merged in? I routinely find my Mac's battery drained while the lid has been closed and the only app preventing sleep is apparently Catch.

**Comment by mipstian** [comment]
The app wouldn't be functional with the PR in this state (see my comments above). If you turned off this new setting, App Nap would kick in and Catch wouldn't be able to check even if the Mac was awake.

I really don't have time these days, so I guess it's up to @jbouaziz or whoever wants to help :)

**Comment by mipstian** [comment]
I implemented this on master, based on this code. Thanks!

**Pull request closed without merging**